lib/SILOptimizer/Utils/PartialApplyCombiner.cpp

```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
   // A partial_apply [stack]'s argument are not owned by the partial_apply and
   // therefore their lifetime must outlive any uses.
-  if (pai->isOnStack()) {
+  if (pai->isOnStack())
     return true;
-  }
 
-  // Copy the original arguments of the partial_apply into newly created
-  // temporaries and use these temporaries instead of the original arguments
-  // afterwards.
-  //
-  // This is done to "extend" the life-time of original partial_apply arguments,
-  // as they may be destroyed/deallocated before the last use by one of the
-  // apply instructions.
-  //
-  // TODO: Copy arguments of the partial_apply into new temporaries only if the
-  // lifetime of arguments ends before their uses by apply instructions.
-  bool needsDestroys = false;
-  CanSILFunctionType paiTy =
-      pai->getCallee()->getType().getAs<SILFunctionType>();
-
-  // Emit a destroy value for each captured closure argument.
-  ArrayRef<SILParameterInfo> paramList = paiTy->getParameters();
-  auto argList = pai->getArguments();
-  paramList = paramList.drop_front(paramList.size() - argList.size());
-
-  llvm::SmallVector<std::pair<SILValue, uint16_t>, 8> argsToHandle;
-  for (unsigned i : indices(argList)) {
-    SILValue arg = argList[i];
-    SILParameterInfo param = paramList[i];
-    // If we do not have an indirect parameter, we do not need to do any further
-    // work here.
-    if (!pai->getSubstCalleeConv().isSILIndirect(param))
-      continue;
-
-    // If our indirect parameter is mutating, we can just skip it as well.
-    if (param.isIndirectMutating())
-      continue;
-
-    // If we are consuming an indirect parameter. Bail! We do not support that
-    // today!
-    if (param.isConsumed()) {
-      return false;
-    }
+  SmallVector<Operand *, 8> argsToHandle;
+  getConsumedPartialApplyArgs(pai, argsToHandle,
+                              /*includeTrivialAddrArgs*/ true);
+  if (argsToHandle.empty())
+    return true;
 
-    // Otherwise, we must have a guaranteed parameter. If we don't, bail.
-    if (!param.isGuaranteed()) {
-      return false;
-    }
+  // Compute the set of endpoints, which will be used to insert destroys of
+  // temporaries.
+  SmallVector<SILInstruction *, 16> paiUsers;
 
-    // If the argument has a dependent type, then we can not create a
-    // temporary for it at the beginning of the function, so we must bail.
-    //
-    // TODO: This is because we are inserting alloc_stack at the beginning/end
-    // of functions where the dependent type may not exist yet.
-    if (arg->getType().hasOpenedExistential())
-      return false;
-
-    // If the temporary is non-trivial, we need to destroy it later.
-    if (!arg->getType().isTrivial(*pai->getFunction()))
-      needsDestroys = true;
-    argsToHandle.push_back(std::make_pair(arg, i));
+  // Of course we must inlude all apply instructions which we want to optimize.
+  for (FullApplySite ai : applies) {
+    paiUsers.push_back(ai.getInstruction());
```

Comment:
What about this:

```
transform(applies, std::back_inserter(paiUsers), [](FullApplySite ai) { return ai.getInstruction(); })
```

Reply:
I find the for-loop version easier to understand.

Reply:
Ok.